src/Orleans/Lifecycle/LifecycleObservable.cs

```diff
@@ -0,0 +1,116 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Linq;
+using System.Threading.Tasks;
+using Orleans.Runtime;
+
+namespace Orleans
+{
+    public class LifecycleObservable<TStage> : ILifecycleObservable<TStage>, ILifecycleObserver
+    {
+        private readonly ConcurrentDictionary<object, OrderedObserver> subscribers;
+        private readonly Logger logger;
+        private TStage highStage;
+
+        public LifecycleObservable(Logger logger)
+        {
+            this.logger = logger?.GetLogger(GetType().Name);
+            subscribers = new ConcurrentDictionary<object, OrderedObserver>();
+        }
+
+        public async Task OnStart()
+        {
+            try
+            {
+                foreach (IGrouping<TStage, OrderedObserver> observerGroup in subscribers.Values
```

Comment:
As we talked in person, hot-path is depending on how intense is the grain activation/deactivation, which is not a design, but a runtime thing. In IoT scenarios where you're working with lot of sensors this can be easily a hot-path. What we know that it is slow and knowing a better/faster way I am not considering premature.